File: Leetcode/1261.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type FindElements struct {
  root *TreeNode  
}


func Constructor(root *TreeNode) FindElements {
    // Восстановим значения узлов дерева с помощью BFS метода
    root.Val=0
    node_queue := []*TreeNode{}
	  node_queue = append(node_queue, root)
    for ok := 1; ok != 0; ok = len(node_queue) {
      current_node := node_queue[0]
      x:=current_node.Val
      if current_node.Left != nil {
        current_node.Left.Val=2*x+1
        node_queue = append(node_queue, current_node.Left)
      }
      if current_node.Right != nil {
        current_node.Right.Val=2*x+2
        node_queue = append(node_queue, current_node.Right)
      }
      node_queue = node_queue[1:]
	  }
    var FE FindElements
    FE.root=root
    return FE
}


func (this *FindElements) Find(target int) bool {
    // Проверим, есть ли узел с заданным значением в массиве с помощью BFS метода
    node_queue := []*TreeNode{}
	  node_queue = append(node_queue, this.root)
    for ok := 1; ok != 0; ok = len(node_queue) {
        current_node := node_queue[0]
        if current_node.Val==target{
          return true
        }
        if current_node.Left != nil {
          node_queue = append(node_queue, current_node.Left)
        }
        if current_node.Right != nil {
          node_queue = append(node_queue, current_node.Right)
        }
        node_queue = node_queue[1:]
	  }
    return false
}


/**
 * Your FindElements object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Find(target);
 */

File: Пукаев/1261.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class FindElements {
public:
    TreeNode* node;
  
    FindElements(TreeNode* root) {
        this->node = root;
        stack<TreeNode*> stackNode;
        root->val = 0;
        stackNode.push(root);
        while(!stackNode.empty()){
            if(stackNode.top()->left && stackNode.top()->left->val == -1){
                stackNode.top()->left->val = 2*stackNode.top()->val + 1;
                cout <<  stackNode.top()->left->val << " ";
               
                stackNode.push(stackNode.top()->left);
                //cout << "top stack: " << stackNode.top()->val;
            }else if(stackNode.top()->right && stackNode.top()->right->val == -1){
                stackNode.top()->right->val = 2*stackNode.top()->val + 2;
                cout <<  stackNode.top()->right->val << " ";
                
                stackNode.push(stackNode.top()->right);
                //cout << "top stack: " << stackNode.top()->val;
                
            }else{
                stackNode.pop();
            }
            
        }
       
    }
    
    bool find(int target) {
        stack<TreeNode*> stackNode;
        stackNode.push(node);
        while(!stackNode.empty()){
            TreeNode* currentNode = stackNode.top();
            stackNode.pop();
            if(currentNode->val == target){
                return true;
            }
            if(currentNode->left){
                stackNode.push(currentNode->left);
            }
            
             if(currentNode->right){
                stackNode.push(currentNode->right);
            }
        }
        return false;
    }
};

/**
 * Your FindElements object will be instantiated and called as such:
 * FindElements* obj = new FindElements(root);
 * bool param_1 = obj->find(target);
 */